### PR TITLE
[language] source code map for bytecode MVP

### PR DESCRIPTION
**Pull request by tzakian**
This PR adds a basic functionality for tracking IR source locations in the
bytecode. There is a lot of work left to do here, but this is an "mvp" for this.

# High level idea

### Functions

Every function definition in the module has a couple mappings and data associated with it:

1. a mapping for type parameters: `<type param offset> -> <type param name & source location>`
2. information on the location of the function: `<source location of the whole function>`
3. a mapping for each code offset: `<codeOffset> -> <sourceLocation>`; and for locals
4. a mapping: `localIndex -> (name, sourceLocation)`.

Since a single source location may lead to multiple bytecode instructions, we represent these offsets as a segment tree (using a btree and appropriate range queries). You can read more in the implementation.

Note that for native functions we only store 1, 2, and 3 (but only for the arguments to the function).

### Structs/Resources

Every also has a similar set of mappings:

1. a mapping for type parameters: `<type param offset> -> <type param name & source location>`
2. information on the location of the struct: `<source location of the whole struct>`
3. a mapping for each field of the struct: `<field offset> -> <name & sourceLocation>`

Note that for native structs that we only keep track of 1 and 2. 

#### From Rust
There are two different ways that you can use this. The first, is from within Rust. You first need to create a `SourceMapping` which requires a compiled module (or script) and a source map. After this you can disassemble individual functions or structs, or the whole module. There are also functions for marking/annotating functions and structs with messages -- reporting of these markings is not yet implemented, but they will be shortly.

#### From the command line

If you compile a module or script (and _NOT_ a program) with the compiler, and pass the `--src-map` flag this will output two different files; a `.mv` (compiled binary), and `.mvsm` (source map). e.g. assume we have a module `foo.mvir` in `$HOME`. and we compile (under `language/compiler`) with:

```
cargo run -- --src-map -m $HOME/foo.mvir
```

Then we can disassemble this in a similar manner by running (under `language/compiler/bytecode_source_map`):

```
cargo run -- -b $HOME/foo.mv
```

### Future work

* Adding bytecode pretty printing that will e.g. will in location indices with the location name.
* Add printing of markings (both function and struct) along with their corresponding source code (if the source code is available).

### An Example

Let's say we have this module (ignore any semantic errors):

```
module Foo {
  resource A {
    a: u64,
    b: bool,
  }

  resource B<T> {
    c: T,
    d: bool,
  }

  struct T {
    foo: u64,
    bar: bool,
  }

  struct R<X> {
    ffoo: u64,
    bbar: bool,
  }

  native struct C<T>;
  native struct D;

  native resource E<T>;
  native resource F;

  public foo<T>(): u64 {
    let x: u64;
    let y: u64;
    x = 1;
    y = 2;
    return move(x) + move(y);
  }

  public bar(x: u64) {
    let t: &u64;
    let y: u64;
    y = move(x);
    t = &y;
    return;
  }

  native nat<T>(x: u64): T;

  public flip(b: bool): u64 {
    if (move(b)) {
      return 1;
    } else {
      return 0;
    }
  }
}
```

Then we produce a source map (internally) that looks like this: 

```
ModuleSourceMap {
    module_name: (
        0000000000000000000000000000000000000000000000000000000000000000,
        Identifier(
            "Foo",
        ),
    ),
    struct_map: {
        4: StructSourceMap {
            decl_location: Span {
                start: ByteIndex(205),
                end: ByteIndex(224),
            },
            type_parameters: [
                (
                    Identifier(
                        "T",
                    ),
                    Span {
                        start: ByteIndex(0),
                        end: ByteIndex(0),
                    },
                ),
            ],
            fields: [],
        },
        0: StructSourceMap {
            decl_location: Span {
                start: ByteIndex(15),
                end: ByteIndex(56),
            },
            type_parameters: [],
            fields: [
                (
                    Identifier(
                        "a",
                    ),
                    Span {
                        start: ByteIndex(32),
                        end: ByteIndex(33),
                    },
                ),
                (
                    Identifier(
                        "b",
                    ),
                    Span {
                        start: ByteIndex(44),
                        end: ByteIndex(45),
                    },
                ),
            ],
        },
        2: StructSourceMap {
            decl_location: Span {
                start: ByteIndex(106),
                end: ByteIndex(149),
            },
            type_parameters: [],
            fields: [
                (
                    Identifier(
                        "foo",
                    ),
                    Span {
                        start: ByteIndex(121),
                        end: ByteIndex(124),
                    },
                ),
                (
                    Identifier(
                        "bar",
                    ),
                    Span {
                        start: ByteIndex(135),
                        end: ByteIndex(138),
                    },
                ),
            ],
        },
        6: StructSourceMap {
            decl_location: Span {
                start: ByteIndex(247),
                end: ByteIndex(268),
            },
            type_parameters: [
                (
                    Identifier(
                        "T",
                    ),
                    Span {
                        start: ByteIndex(0),
                        end: ByteIndex(0),
                    },
                ),
            ],
            fields: [],
        },
        7: StructSourceMap {
            decl_location: Span {
                start: ByteIndex(271),
                end: ByteIndex(289),
            },
            type_parameters: [],
            fields: [],
        },
        1: StructSourceMap {
            decl_location: Span {
                start: ByteIndex(60),
                end: ByteIndex(102),
            },
            type_parameters: [
                (
                    Identifier(
                        "T",
                    ),
                    Span {
                        start: ByteIndex(0),
                        end: ByteIndex(0),
                    },
                ),
            ],
            fields: [
                (
                    Identifier(
                        "c",
                    ),
                    Span {
                        start: ByteIndex(80),
                        end: ByteIndex(81),
                    },
                ),
                (
                    Identifier(
                        "d",
                    ),
                    Span {
                        start: ByteIndex(90),
                        end: ByteIndex(91),
                    },
                ),
            ],
        },
        5: StructSourceMap {
            decl_location: Span {
                start: ByteIndex(227),
                end: ByteIndex(243),
            },
            type_parameters: [],
            fields: [],
        },
        3: StructSourceMap {
            decl_location: Span {
                start: ByteIndex(153),
                end: ByteIndex(201),
            },
            type_parameters: [
                (
                    Identifier(
                        "X",
                    ),
                    Span {
                        start: ByteIndex(0),
                        end: ByteIndex(0),
                    },
                ),
            ],
            fields: [
                (
                    Identifier(
                        "ffoo",
                    ),
                    Span {
                        start: ByteIndex(171),
                        end: ByteIndex(175),
                    },
                ),
                (
                    Identifier(
                        "bbar",
                    ),
                    Span {
                        start: ByteIndex(186),
                        end: ByteIndex(190),
                    },
                ),
            ],
        },
    },
    function_map: {
        2: FunctionSourceMap {
            decl_location: Span {
                start: ByteIndex(509),
                end: ByteIndex(534),
            },
            type_parameters: [
                (
                    Identifier(
                        "T",
                    ),
                    Span {
                        start: ByteIndex(0),
                        end: ByteIndex(0),
                    },
                ),
            ],
            locals: [
                (
                    Identifier(
                        "x",
                    ),
                    Span {
                        start: ByteIndex(523),
                        end: ByteIndex(524),
                    },
                ),
            ],
            code_map: {},
        },
        3: FunctionSourceMap {
            decl_location: Span {
                start: ByteIndex(538),
                end: ByteIndex(639),
            },
            type_parameters: [],
            locals: [
                (
                    Identifier(
                        "b",
                    ),
                    Span {
                        start: ByteIndex(550),
                        end: ByteIndex(551),
                    },
                ),
            ],
            code_map: {
                0: Span {
                    start: ByteIndex(574),
                    end: ByteIndex(581),
                },
                2: Span {
                    start: ByteIndex(598),
                    end: ByteIndex(599),
                },
                3: Span {
                    start: ByteIndex(591),
                    end: ByteIndex(599),
                },
                4: Span {
                    start: ByteIndex(627),
                    end: ByteIndex(628),
                },
                5: Span {
                    start: ByteIndex(620),
                    end: ByteIndex(628),
                },
            },
        },
        0: FunctionSourceMap {
            decl_location: Span {
                start: ByteIndex(293),
                end: ByteIndex(403),
            },
            type_parameters: [
                (
                    Identifier(
                        "T",
                    ),
                    Span {
                        start: ByteIndex(0),
                        end: ByteIndex(0),
                    },
                ),
            ],
            locals: [
                (
                    Identifier(
                        "x",
                    ),
                    Span {
                        start: ByteIndex(324),
                        end: ByteIndex(325),
                    },
                ),
                (
                    Identifier(
                        "y",
                    ),
                    Span {
                        start: ByteIndex(340),
                        end: ByteIndex(341),
                    },
                ),
            ],
            code_map: {
                0: Span {
                    start: ByteIndex(356),
                    end: ByteIndex(357),
                },
                1: Span {
                    start: ByteIndex(352),
                    end: ByteIndex(353),
                },
                2: Span {
                    start: ByteIndex(367),
                    end: ByteIndex(368),
                },
                3: Span {
                    start: ByteIndex(363),
                    end: ByteIndex(364),
                },
                4: Span {
                    start: ByteIndex(381),
                    end: ByteIndex(388),
                },
                5: Span {
                    start: ByteIndex(391),
                    end: ByteIndex(398),
                },
                6: Span {
                    start: ByteIndex(381),
                    end: ByteIndex(398),
                },
                7: Span {
                    start: ByteIndex(374),
                    end: ByteIndex(398),
                },
            },
        },
        1: FunctionSourceMap {
            decl_location: Span {
                start: ByteIndex(407),
                end: ByteIndex(505),
            },
            type_parameters: [],
            locals: [
                (
                    Identifier(
                        "x",
                    ),
                    Span {
                        start: ByteIndex(418),
                        end: ByteIndex(419),
                    },
                ),
                (
                    Identifier(
                        "t",
                    ),
                    Span {
                        start: ByteIndex(436),
                        end: ByteIndex(437),
                    },
                ),
                (
                    Identifier(
                        "y",
                    ),
                    Span {
                        start: ByteIndex(453),
                        end: ByteIndex(454),
                    },
                ),
            ],
            code_map: {
                0: Span {
                    start: ByteIndex(469),
                    end: ByteIndex(476),
                },
                1: Span {
                    start: ByteIndex(465),
                    end: ByteIndex(466),
                },
                2: Span {
                    start: ByteIndex(486),
                    end: ByteIndex(488),
                },
                3: Span {
                    start: ByteIndex(482),
                    end: ByteIndex(483),
                },
                4: Span {
                    start: ByteIndex(494),
                    end: ByteIndex(500),
                },
            },
        },
    },
}
```

and will dissasemble through the CLI as:
```
module 0000000000000000000000000000000000000000000000000000000000000000.Foo {
resource A<> {
	a: u64,
	b: bool
}
resource B<T: All> {
	c: T,
	d: bool
}
struct T<> {
	foo: u64,
	bar: bool
}
struct R<X: All> {
	ffoo: u64,
	bbar: bool
}
native struct C<T: All> {

}
native struct D<> {

}
native resource E<T: All> {

}
native resource F<> {

}

public foo<T: All>():u64
public bar<>(x: u64):
native nat<T: All>(x: u64):T
public flip<>(b: bool):u64
}
```

### CODE REVIEW

File: language/compiler/bytecode_source_map/src/source_map.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+use std::ops::Bound;
+use types::account_address::AccountAddress;
+use types::identifier::Identifier;
+use vm::file_format::{CodeOffset, FunctionDefinitionIndex, TableIndex};
+
+//***************************************************************************
+// Source location mapping
+//***************************************************************************
+
+pub type SourceMap = Vec<ModuleSourceMap>;
```
Comment (sblackshear) [inline language/compiler/bytecode_source_map/src/source_map.rs]: The only place I see this type used is in `pub fn source_map_from_file(filename: &str) -> SourceMap`.

Would we want to store more than one `ModuleSourceMap` in a single file? It seems like one map per file (+ a predictable naming convention for the file) might make it easier to use and update source maps produced by by multiple runs of the compiler. For example: imagine you run `movec M1.mvir M2.mvir M3.mvir` and get one file with three maps, then change `M1.mvir` and re-run `movec M1.mvir` to get one file with one map. It wouldn't be clear whether to use the `M1` source map from the first or second run.

By contrast, if you get one source map file per module, the old `M1` source map would get clobbered by the second invocation and there would be no confusion.

File: language/stdlib/src/lib.rs [hunk 2]
```diff
@@ -19,18 +20,27 @@ lazy_static! {
...
 pub fn stdlib_modules() -> &'static [VerifiedModule] {
-    &*STDLIB_MODULES
+    &*ANNOTATED_STDLIB.0
+}
+
+/// Returns a reference to the source maps for the standard library.
+///
+/// The order of the modules returned in this follows the same order as the modules returned in the
+/// stdlib_modules.
+pub fn stdlib_source_map() -> &'static [ModuleSourceMap] {
```
Comment (sblackshear) [inline language/stdlib/src/lib.rs]: Is there Rust code that would want to call this?

File: language/compiler/ir_to_bytecode/src/context.rs [hunk 3]
```diff
@@ -289,6 +299,33 @@ impl<'a> Context<'a> {
...
+
+    pub fn incr_function_index(&mut self) {
+        match self.current_function_index {
+            FunctionDefinitionIndex(i) => {
+                self.current_function_index = FunctionDefinitionIndex(i + 1)
+            }
+        }
+    }
+
+    pub fn current_function_index(&self) -> FunctionDefinitionIndex {
```
Comment (sblackshear) [inline language/compiler/ir_to_bytecode/src/context.rs]: Is this unused? I don't see any code that calls it. What would happen if we deleted `current_function_index`?

Comment (tzakian) [reply language/compiler/ir_to_bytecode/src/context.rs]: Agh! That was on the chopping block, and I meant to remove it. (Like I said, cleanup is needed....). If you remove it, nothing will break right now.

File: language/compiler/bytecode_source_map/src/source_map.rs (hunk 1 above)
Comment (tzakian) [reply language/compiler/bytecode_source_map/src/source_map.rs]: You do only get one `ModuleSourceMap` per module. However, `CompiledPrograms` can lead to multiple modules being compiled which unless I'm misunderstanding something means that we can have a many-module to single-file mapping. But I may be misunderstanding how programs are represented (and I couldn't find an example readily). 

I agree, if we can have a 1-1 mapping of file to module that would be the best, but I'm not certain that that's possible.  I'll take a look at this tomorrow though and see if we can have this 1-1 setting, because I think this will definitely be the best.

File: language/stdlib/src/lib.rs (hunk 2 above)
Comment (tzakian) [reply language/stdlib/src/lib.rs]: Not currently, but my thinking is that there would be (e.g. Clark's code). And, having these prebuilt for the stdlib makes sense since the stdlib is already compiled as well.

File: language/compiler/bytecode_source_map/src/source_map.rs (hunk 1 above)
Comment (sblackshear) [reply language/compiler/bytecode_source_map/src/source_map.rs]: Your understanding is totally correct given the state of the code. But as @runtian-zhou mentioned above, we are deprecating `CompiledProgram` and moving to a bytecode file format where there is exactly one module *or* (not and) script per file. So I think we can have that 1-1 mapping!

File: language/compiler/bytecode_source_map/src/source_map.rs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+    pub code_map: BTreeMap<CodeOffset, Loc>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ModuleSourceMap {
+    // The name <address.module_name> for module that this source map is for
+    pub module_name: (AccountAddress, Identifier),
+
+    // A mapping of StructDefinitionIndex to source map for each struct/resource
+    struct_map: HashMap<TableIndex, StructSourceMap>,
```
Comment (sblackshear) [inline language/compiler/bytecode_source_map/src/source_map.rs]: Nit: can we make these into `BTreeMap`s? If someone iterated over the marks, they could get different orderings in different iters. This might or might not matter, but seems good to avoid the surpris if we can.

File: language/compiler/bytecode_source_map/src/source_map.rs [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+        self.type_parameters.push(type_name)
+    }
+
+    pub fn get_type_parameter_name(&self, type_parameter_idx: usize) -> Option<SourceName> {
+        self.type_parameters.get(type_parameter_idx).cloned()
+    }
+
+    /// A single source-level instruction may possibly map to a number of bytecode instructions. In
+    /// order to not store a location for each instruction, we instead use a BTreeMap to represent
+    /// a segement map (holding the left-hand-sands of the segments).  Thus, an instruction
```
Comment (sblackshear) [inline language/compiler/bytecode_source_map/src/source_map.rs]:
```suggestion
    /// a segment map (holding the left-hand-sides of the segments). Thus, an instruction
```

File: language/compiler/bytecode_source_map/src/source_map.rs [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    /// sequence is always marked from the starting point. To determine what part of the source
+    /// code corresponds to a given `CodeOffset` we query to find the element that is the largest
+    /// number less than or equal to the query. This will give us the location for that bytcode
```
Comment (sblackshear) [inline language/compiler/bytecode_source_map/src/source_map.rs]:
```suggestion
    /// number less than or equal to the query. This will give us the location for that bytecode
```

File: language/compiler/bytecode_source_map/src/source_map.rs [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    /// range.
+    pub fn add_code_mapping(&mut self, start_offset: CodeOffset, location: Loc) {
+        self.code_map.insert(start_offset, location);
+    }
+
+    // NB that locations must be added in order.
```
Comment (sblackshear) [inline language/compiler/bytecode_source_map/src/source_map.rs]: What is NB?

File: language/compiler/bytecode_source_map/src/mapping.rs [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+    CodeOffset, CompiledModule, CompiledScript, FieldDefinitionIndex, FunctionDefinition,
+    FunctionDefinitionIndex, SignatureToken, StructDefinition, StructDefinitionIndex,
+    StructFieldInformation, TableIndex, TypeSignature,
+};
+
+/// An object that associates source code with compiled bytecode and source map.
+#[derive(Debug)]
+pub struct SourceMapping {
+    // The various options that we can set for disassembly.
+    options: SourceMappingOptions,
```
Comment (sblackshear) [inline language/compiler/bytecode_source_map/src/mapping.rs]: Do we need these both here and in the disassembler `main`? It seems like having the options live only in the `main` and having the caller pass them into the disassemble logic would be more functional/easier to understand.

File: language/compiler/bytecode_source_map/src/mapping.rs [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+    _source_code: Option<String>,
+
+    // The source map for the bytecode made w.r.t. to the `source_code`
+    source_map: ModuleSourceMap,
+
+    // The resulting bytecode from compiling the source map
+    bytecode: CompiledModule,
+
+    // Any function markings
+    function_marks: HashMap<TableIndex, FunctionMarking>,
```
Comment (sblackshear) [inline language/compiler/bytecode_source_map/src/mapping.rs]: It might might sense to have an immutable `SourceMapping` that associates bytecode with the source + source map and having a separate `MarkedSourceMapping` type that wraps an immutable source mapping + can be mutated (but only to touch the markings).

File: language/compiler/bytecode_source_map/src/mapping.rs [hunk 10]
```diff
@@ -0,0 +1,486 @@
...
+            .iter()
+            .zip(function_source_map.locals.iter())
+            .map(|(typ, (name, _))| {
+                let ty_str =
+                    self.disassemble_sig_tok(typ.clone(), &function_source_map.type_parameters)?;
+                Ok(format!("{}: {}", name.to_string(), ty_str))
+            })
+            .collect::<Result<Vec<String>>>()?;
+        Ok(format!(
+            "{visibility_modifier}{name}<{ty_params}>({args}):{ret_type}",
```
Comment (sblackshear) [inline language/compiler/bytecode_source_map/src/mapping.rs]: Would adding the `<>` around `ty_params` above only in the case where they're non-empty allow you to prevent the funky `<>` printing?

File: language/compiler/bytecode_source_map/src/mapping.rs [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+                .map(|(name, ty)| {
+                    let ty_str =
+                        self.disassemble_sig_tok(ty.0.clone(), &struct_source_map.type_parameters)?;
+                    Ok(format!("{}: {}", name.to_string(), ty_str))
+                })
+                .collect::<Result<Vec<String>>>()?,
+        };
+
+        Ok(format!(
+            "{native}{nominal_name} {name}<{ty_params}> {{\n\t{fields}\n}}",
```
Comment (sblackshear) [inline language/compiler/bytecode_source_map/src/mapping.rs]: Similar question here.

File: language/compiler/bytecode_source_map/src/mapping.rs (hunk 8 above)
Comment (tzakian) [reply language/compiler/bytecode_source_map/src/mapping.rs]: I wanted to decouple the `main` from the implementation of the disassembler -- since we also want to support calling this from within Rust itself.

File: language/compiler/bytecode_source_map/src/source_map.rs (hunk 7 above)
Comment (tzakian) [reply language/compiler/bytecode_source_map/src/source_map.rs]: "Nota Bene" or note well (https://en.wikipedia.org/wiki/Nota_bene). I'll go through and update with something more readable.

File: language/compiler/bytecode_source_map/src/mapping.rs (hunk 10 above)
Comment (tzakian) [reply language/compiler/bytecode_source_map/src/mapping.rs]: yes! I was thinking of doing that along with not printing '{' and '}' for empty (i.e. native) structs.

File: language/compiler/bytecode_source_map/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    /// directory (.mvir, .mvsm, and .mv).
+    #[structopt(short = "s", long = "script")]
+    pub is_script: bool,
+
+    /// The path to the bytecode file. We assume that all three files that we need reside under this
+    /// directory (.mvir, .mvsm, and .mv).
+    #[structopt(short = "b", long = "bytecode")]
+    pub bytecode_file_path: String,
+
+    /// Output file, stdout if not present.
```
[the remark this answers is not in the document]
Comment (tzakian) [reply language/compiler/bytecode_source_map/src/main.rs]: Fair enough. It also makes the program smaller 💯

File: language/compiler/bytecode_source_map/src/mapping.rs (hunk 9 above)
Comment (tzakian) [reply language/compiler/bytecode_source_map/src/mapping.rs]: I like this idea. There's still a lot of work left in marking and printing marked functions. How about I add this in a separate diff?

File: language/compiler/ir_to_bytecode/src/context.rs [hunk 13]
```diff
@@ -289,6 +299,93 @@ impl<'a> Context<'a> {
         Ok(())
     }
 
+    //**********************************************************************************************
+    // Source Maps
+    //**********************************************************************************************
+
+    pub fn add_top_level_function_mapping(
```
Comment (tnowacki) [inline language/compiler/ir_to_bytecode/src/context.rs]: Can you just make the source map a public field? Then you don't need all these delegates

**Comment by tzakian** [comment]
FYI: Changing some things around, so don't review this yet.

File: language/compiler/ir_to_bytecode/src/context.rs [hunk 14]
```diff
@@ -289,6 +299,93 @@ impl<'a> Context<'a> {
...
+        &mut self,
+        struct_def_idx: StructDefinitionIndex,
+        location: Loc,
+    ) -> Result<()> {
+        self.source_map
+            .add_struct_field_mapping(struct_def_idx, location)
+    }
+
+    pub fn incr_function_index(&mut self) {
+        match self.current_function_index {
```
Comment (tnowacki) [inline language/compiler/ir_to_bytecode/src/context.rs]:
```
let FunctionDefinitionIndex(i) = self.current_function_index;
self.current_function_index = FunctionDefinitionIndex(i + 1)
```

File: language/compiler/ir_to_bytecode/src/context.rs [hunk 15]
```diff
@@ -289,6 +299,93 @@ impl<'a> Context<'a> {
...
+    pub fn add_field_mapping(
+        &mut self,
+        struct_def_idx: StructDefinitionIndex,
+        location: Loc,
+    ) -> Result<()> {
+        self.source_map
+            .add_struct_field_mapping(struct_def_idx, location)
+    }
+
+    pub fn incr_function_index(&mut self) {
```
Comment (tnowacki) [inline language/compiler/ir_to_bytecode/src/context.rs]: I'd probably prefer a `set` function than an `incr` function

`incr` feels brittle to me

File: language/compiler/ir_to_bytecode/src/compiler.rs [hunk 16]
```diff
@@ -26,13 +27,56 @@ use std::{
...
+macro_rules! record_srcloc_info {
+    ($context:expr, $code:expr, $loc:expr) => {
+        let code_offset = $code.len() as CodeOffset;
+        $context.add_source_info(
+            $context.current_function_definition_index(),
+            code_offset,
+            $loc,
+        )?;
+    };
+    (local | $context:expr, $var:expr) => {
```
Comment (tnowacki) [inline language/compiler/ir_to_bytecode/src/compiler.rs]: Not sure if I really like the `|` syntax.

Thoughts on `(local: `?

File: language/compiler/ir_to_bytecode/src/context.rs (hunk 13 above)
Comment (tnowacki) [reply language/compiler/ir_to_bytecode/src/context.rs]: Or you can make the map not part of the context, whatever is easier. But I'd prefer no delegates :)

File: language/compiler/ir_to_bytecode/src/compiler.rs [hunk 17]
```diff
@@ -26,13 +27,56 @@ use std::{
...
     file_format::{
-        self, Bytecode, CodeUnit, CompiledModule, CompiledModuleMut, CompiledProgram,
+        self, Bytecode, CodeOffset, CodeUnit, CompiledModule, CompiledModuleMut, CompiledProgram,
         CompiledScript, CompiledScriptMut, FieldDefinition, FieldDefinitionIndex,
         FunctionDefinition, FunctionSignature, Kind, LocalsSignature, MemberCount, SignatureToken,
         StructDefinition, StructFieldInformation, StructHandleIndex, TableIndex,
     },
 };
 
+macro_rules! record_srcloc_info {
```
Comment (tnowacki) [inline language/compiler/ir_to_bytecode/src/compiler.rs]: This feels a bit wordy to me. Thoughts on `record_loc` ? or `record_src_loc`? 

Just `srcloc_info` feels a bit redundant?

File: language/compiler/bytecode_source_map/src/utils.rs [hunk 18]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::source_map::{ModuleSourceMap, SourceMap};
+use std::fs::File;
+use std::path::Path;
+
+pub fn module_source_map_from_file(file_path: &Path) -> ModuleSourceMap {
+    let file = File::open(file_path).expect("Failed to open source map file");
+    serde_json::from_reader(file).expect("Error while reading in source map information")
```
Comment (tnowacki) [inline language/compiler/bytecode_source_map/src/utils.rs]: Magic!

File: language/compiler/bytecode_source_map/src/source_map.rs [hunk 19]
```diff
@@ -0,0 +1,319 @@
...
+// Source location mapping
+//***************************************************************************
+
+pub type SourceMap = Vec<ModuleSourceMap>;
+pub type SourceName = (Identifier, Loc);
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct StructSourceMap {
+    /// The source declaration location of the struct
+    pub decl_location: Loc,
```
Comment (tnowacki) [inline language/compiler/bytecode_source_map/src/source_map.rs]: Does Rust have functors? :p Otherwise can we parameterize this by the type of Loc? 

I can make a separate PR for that if you want, but I'll need it for the source language source information (as its spread across multiple files so doesn't line up with the IR)

File: language/compiler/ir_to_bytecode/src/compiler.rs [hunk 20]
```diff
@@ -797,6 +867,7 @@ fn compile_command(
             compile_expression(context, function_frame, code, *e)?;
 
             let def_idx = context.struct_definition_index(&name)?;
+            record_srcloc_info!(context, code, cmd.span);
```
Comment (tnowacki) [inline language/compiler/ir_to_bytecode/src/compiler.rs]: we need to record a loc everytime we push right? 

Maybe make a macro for that?

File: language/compiler/ir_to_bytecode/src/context.rs (hunk 14 above)
Comment (tnowacki) [reply language/compiler/ir_to_bytecode/src/context.rs]: Or 
```
self.current_function_index = FunctionDefinitionIndex(self.current_function_index.0 + 1)
```

File: language/compiler/ir_to_bytecode/src/compiler.rs (hunk 20 above)
Comment (tzakian) [reply language/compiler/ir_to_bytecode/src/compiler.rs]: We actually don't!

As an example take an `if` expression. We want the `Branch` for this to be associated with the consequent branch of source-level expression. Thus, there is a many-to-one mapping of bytecode offset to source location. So having this isn't as useful IMO.

File: language/compiler/ir_to_bytecode/src/compiler.rs [hunk 21]
```diff
@@ -26,13 +27,74 @@ use std::{
...
+    (struct_decl: $context:expr, $location:expr) => {
+        $context
+            .source_map
+            .add_top_level_struct_mapping($context.current_struct_definition_index(), $location)?;
+    };
+}
+
+macro_rules! make_push_instr {
+    ($context:ident, $code:ident) => {
+        macro_rules! push_instr {
```
Comment (tnowacki) [inline language/compiler/ir_to_bytecode/src/compiler.rs]: Glorious

File: language/compiler/bytecode_source_map/src/source_map.rs [hunk 22]
```diff
@@ -0,0 +1,326 @@
...
+    /// A single source-level instruction may possibly map to a number of bytecode instructions. In
+    /// order to not store a location for each instruction, we instead use a BTreeMap to represent
+    /// a segment map (holding the left-hand-sides of each segment).  Thus, an instruction
+    /// sequence is always marked from its starting point. To determine what part of the source
+    /// code corresponds to a given `CodeOffset` we query to find the element that is the largest
+    /// number less than or equal to the query. This will give us the location for that bytecode
+    /// range.
+    pub fn add_code_mapping(&mut self, start_offset: CodeOffset, location: Location) {
+        let possible_segment = self.get_code_location(start_offset);
+        if possible_segment.is_none() || possible_segment.unwrap() != location {
```
Comment (tnowacki) [inline language/compiler/bytecode_source_map/src/source_map.rs]:
```
match possible_segment_opt {
  Some(posssible_segment) if possible_sgment != location => { self.code_map.insert(start_offset, location); }
  _ => ()
}
```
